src/scripting_api/event.js

```diff
@@ -47,11 +47,32 @@ class EventDispatcher {
     this._document.obj._eventDispatcher = this;
   }
 
+  /*
+   * Take an event object and reconstruct what we think the result will be once
+   * the change is applied, so we can validate against it and cancel the event
+   * if need be
+   *
+   * TODO: Given the event info we currently have, we can't determine where in
+   * the value a single-character insertion/deletion happened.  For now we just
+   * assume they happen at the end of the string, which is by far the most
+   * common case, but this leaves some edge cases, see issue #14307.
+   */
   mergeChange(event) {
     let value = event.value;
     if (typeof value !== "string") {
       value = value.toString();
     }
+    // If there was no selection, it's a single-character change
```

Comment:
Where did you find this information ?
In the specs (see http://dev.datalogics.com/cookbook/document/AcrobatDC_js_api_reference.pdf#page=381&zoom=auto,-214,751) about `selStart`:
`The starting position of the current text selection during a keystroke event.`
I just tried in Acrobat to add text in textfield and print the changes.
 - I entered: `abcdef`
 - I selected `bcd` and I pressed `g`
and I get the following output:
```
change=a, selStart=0, selEnd=0
change=b, selStart=1, selEnd=1
change=c, selStart=2, selEnd=2
change=d, selStart=3, selEnd=3
change=e, selStart=4, selEnd=4
change=f, selStart=5, selEnd=5
change=g, selStart=1, selEnd=4
```
My understanding is that the selStart is the position of the caret before the change.
When there is no selection, selEnd is equals to selStart else it's the position of the end of the selection.

Reply:
I found it through debugging and seeing what the events actually looked like - like I said (maybe here or on the bug, I don't remember), I wasn't sure if how these events happen was specified in a doc somewhere or not - it looks like it is, which I guess means the bug is that pdf.js isn't populating selStart/selEnd for these events? That would also make sense, and would solve the edge cases as well.